### PR TITLE
fix: 1554 lag correctness

### LINKED ISSUE #1554: `lag()/at()/lead()` return show offset'th row, it is not related to window frame bound
**Bug Description**

```yaml
  - id: 6
    desc: window merge optimization
    inputs:
      - columns: [ "row_id int","ts timestamp","group1 string","val1 int" ]
        indexs: [ "index1:group1:ts" ]
        name: t1
        data: |
          1, 1612130400000, g1, 1
          2, 1612130401000, g1, 2
          3, 1612130402000, g1, 3
          4, 1612130403000, g1, 4
          5, 1612130404000, g1, 5
    sql: |
      select
      `row_id` as row_id_1,
      `row_id` as t1_row_id_original_0,
      case when !isnull(at(`val1`, 0)) over t1_group1_ts_0s_5s_10 then count_where(`val1`, `val1` = at(`val1`, 0)) over t1_group1_ts_0s_5s_10 else null end as t1_val1_window_count_1,
      case when !isnull(at(`val1`, 0)) over t1_group1_ts_1s_5s_10 then count_where(`val1`, `val1` = at(`val1`, 0)) over t1_group1_ts_1s_5s_10 else null end as t1_val1_window_count_2
      from
        `t1` WINDOW
        t1_group1_ts_0s_5s_10 as (partition by `group1` order by `ts` rows_range between 5s preceding and 0s preceding MAXSIZE 10),
        t1_group1_ts_1s_5s_10 as (partition by `group1` order by `ts` rows_range between 5s preceding and 1s preceding MAXSIZE 10);
    expect:
      columns: ["row_id_1 int", "t1_row_id_original_0 int", "t1_val1_window_count_1 int64", "t1_val1_window_count_2 int64"]
      order: row_id_1
      data: |
        1, 1, 1, 0
        2, 2, 1, 0
        3, 3, 1, 0
        4, 4, 1, 0
        5, 5, 1, 0
```

**Expected Behavior**

the case should pass. Current result is not:

```log
+----------+----------------------+------------------------+------------------------+
| row_id_1 | t1_row_id_original_0 | t1_val1_window_count_1 | t1_val1_window_count_2 |
+----------+----------------------+------------------------+------------------------+
| 1        | 1                    | 1                      | NULL                   |
| 2        | 2                    | 1                      | NULL                   |
| 3        | 3                    | 1                      | NULL                   |
| 4        | 4                    | 1                      | NULL                   |
| 5        | 5                    | 1                      | NULL                   |
+----------+----------------------+------------------------+------------------------+
```

**Work List**
- [x] add extra handling in logic plan for `lag` functions
  - working in #1605 
- [x] fix lag correctness in request mode & cluster environment
- [x] fix window merge result correctness ( may related to #1587 )

### PR DESCRIPTION
**Pull request by aceforeverd**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

close #1554 

for those where `lag` function exists in sql, e.g
```sql
select id, sum(col) over w1, lag(col, 0) over w1 from t1 window w1 as (....)
```

the logic plan is constructed,  previous:

```
Query
  ProjectPlan
    project_vec
      -ProjectList 1
        - id
        - sum(col)
        - lag(col, 0)
````

after the PR:

add a restriction to `lag`'s offset: it must be constant

the `lag` project node will splited into seperate project list if:
- `lag` over a `rows_range` window

new window is a `ROWS` type window:
- without `max_size`
- window start = offset
- window end = current row

This is a internal split that happens before window merge, so mergable window can still merged based on original logic

### CODE REVIEW

**Comment by aceforeverd** [comment]
[High-priority Bugs](https://trello.com/c/h7opm5kE/45-high-priority-bugs)

**Comment by aceforeverd** [comment]
TODO: discuss TTL cal in deploy progress

File: hybridse/src/plan/planner.cc [hunk 1]
```diff
@@ -1118,6 +1134,136 @@ base::Status Planner::TransformTableDef(const std::string &table_name, const Nod
...
+
+            bool preserve_old_window = !window_bound_relative_projects->GetProjects().empty();
+
+            if (preserve_old_window) {
+                splited_map.emplace(kv.first, window_bound_relative_projects);
+            }
+
+            auto* window_bound_non_relative_window = CreateWindowUnboundPrecedingAndCurrent(kv.first);
+            // prohibit window merge on the window
+            window_bound_non_relative_window->SetPermitWindowMerge(false);
```
Comment (zhanghaohit) [inline hybridse/src/plan/planner.cc]: maybe we can verify whether this is necessary. Further merging should work properly.

File: hybridse/src/planv2/planner_v2_test.cc [hunk 2]
```diff
@@ -331,50 +331,45 @@ TEST_F(PlannerV2Test, MultiProjectListPlanPostTest) {
...
+        R"sql(SELECT
+        sum(col1) OVER w1 as w1_col1_sum,
+        sum(col3) OVER w2 as w2_col3_sum,
+        sum(col4) OVER w2 as w2_col4_sum,
+        col1,
+        sum(col3) OVER w1 as w1_col3_sum,
+        col2,
+        sum(col1) OVER w2 as w2_col1_sum,
+        lag(col1, 0) OVER w2 as w2_col1_at_0,
+        lag(col1, 1) OVER w2 as w2_col1_at_1
```
Comment (zhanghaohit) [inline hybridse/src/planv2/planner_v2_test.cc:349]: maybe we can move one `lag` to the middle of the projection list to check the ordering is correct

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Comment by aceforeverd** [comment]
Don't merge util performance issue resolved

**Review by zhanghaohit — DISMISSED** [review]
performance issue of unbound window frame

File: hybridse/src/planv2/planner_v2_test.cc (hunk 2 above)
Comment (aceforeverd) [reply hybridse/src/planv2/planner_v2_test.cc:349]: added

File: hybridse/src/plan/planner.cc (hunk 1 above)
Comment (aceforeverd) [reply hybridse/src/plan/planner.cc]: yeah, I removed this

File: hybridse/src/plan/planner.h [hunk 3]
```diff
@@ -74,20 +74,36 @@ class Planner {
...
+    const bool enable_batch_window_parallelization_;
+    node::NodeManager *node_manager_;
+
+ private:
+    // get the `WindowDefNode` for `lag(col, offset)` function
+    //  output a rows window, where
+    //  - frame_start = max(offset, in.frame_start)
+    //  - frame_end = current row
+    absl::StatusOr<node::WindowDefNode *> ConstructWindowForLag(const node::WindowDefNode *in,
+                                                             const node::CallExprNode *call) const;
```
Comment (aceforeverd) [inline hybridse/src/plan/planner.h]: fmt

File: hybridse/src/plan/planner.cc [hunk 4]
```diff
@@ -1128,6 +1162,56 @@ base::Status Planner::TransformTableDef(const std::string &table_name, const Nod
...
+    }
+
+    auto *const_node = dynamic_cast<node::ConstNode *>(offset_expr);
+    int64_t offset = const_node->GetAsInt32();
+    switch (in->GetFrame()->frame_type()) {
+        case node::FrameType::kFrameRows: {
+            int64_t win_frame_start_offset = in->GetFrame()->GetHistoryRowsStartPreceding();
+
+            auto *rows_frame_ext = node_manager_->MakeFrameExtent(
+                node_manager_->MakeFrameBound(node::BoundType::kPreceding, std::max(offset, win_frame_start_offset)),
```
Comment (aceforeverd) [inline hybridse/src/plan/planner.cc]: offfset

File: cases/function/function/test_udaf_function.yaml [hunk 5]
```diff
@@ -2332,24 +2331,182 @@ cases:
...
+        5, 5, 5, 4,    2
+        6, 4, 4, NULL, NULL
+        7, 3, 3, 4,    NULL
+        8, 2, 2, 3,    NULL
 
+  - id: 58
+    desc: |
+      correctness for at/lag when offset out-of-range rows_range window frame bound, together with other window function.
+      refer https://github.com/4paradigm/OpenMLDB/issues/1554
+    # FIXME(ace): request/cluster run do not meet the expects
```
Comment (aceforeverd) [inline cases/function/function/test_udaf_function.yaml]: will fix after #1739 , merge #1739 first

File: hybridse/src/plan/planner.cc [hunk 6]
```diff
@@ -1128,6 +1162,42 @@ base::Status Planner::TransformTableDef(const std::string &table_name, const Nod
...
+        return absl::InvalidArgumentError("offset can only be constant");
+    }
+
+    if (in->GetFrame()->frame_type() != node::FrameType::kFrameRows &&
+        in->GetFrame()->frame_type() != node::FrameType::kFrameRowsRange) {
+        return absl::InvalidArgumentError("input window is not a ROWS or ROWS_RANGE window");
+    }
+
+    auto *const_node = dynamic_cast<node::ConstNode *>(offset_expr);
+    int64_t offset = const_node->GetAsInt32();
```
Comment (zhanghaohit) [inline hybridse/src/plan/planner.cc]: why not use `GetAsInt64()` directly?

**Review by zhanghaohit — APPROVED** [review]

File: hybridse/src/plan/planner.cc (hunk 6 above)
Comment (aceforeverd) [reply hybridse/src/plan/planner.cc]: resolved

**Pull request merged**